### PR TITLE
Support for Hugging Face model finetuning.

### PR DESCRIPTION
**Pull request by albertogaspar**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Models from `transformers` cannot be used within Fortuna.

## What is the new behavior?

Models from `transformers` can now be finetuned within Fortuna

- added possibility to have gradient clipping
- added possibility to have gradient accumulation
- added support for mixed precision using Flax DynamicScale

### CODE REVIEW

File: fortuna/utils/optimizer.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    A common ImageNet schedule decays the learning rate by a factor of 0.1 at epochs 30, 60 and 80.
+    This would be specified setting `boundaries_and_scales` to:
+
+      {
+        30: 0.1,
+        60: 0.01,
+        80: 0.001
+      }
+
+    Args:
```
Comment (gianlucadetommaso) [inline fortuna/utils/optimizer.py:29]: Parameters? Adjust formatting.

File: fortuna/utils/training.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    clip_fn = lambda z: jnp.clip(z, -max_grad_val, max_grad_val)
+    grad = tree_map(clip_fn, grad)
+    return grad
+
+
+def clip_grandients_by_norm(grad: jnp.ndarray, max_grad_norm: float) -> jnp.ndarray:
+    grad_norm = jnp.sqrt(
+        tree_reduce(
+            lambda x, y: x + jnp.sum(y ** 2), grad, initializer=0
+        )
```
Comment (gianlucadetommaso) [inline fortuna/utils/training.py:15]: Nice using `tree_reduce` for this, I'll have to remember it.

**Review by gianlucadetommaso — COMMENTED** [review]
Just a few comments, I think this is almost ready to be merged! Great work 😄

File: fortuna/model/model_manager/classification.py [hunk 3]
```diff
@@ -238,3 +262,83 @@ def init(
...
+        self,
+        params: Params,
+        inputs: Dict[str, Array],
+        mutable: Optional[Mutable] = None,
+        train: bool = False,
+        rng: Optional[PRNGKeyArray] = None,
+        **kwargs,
+    ) -> Union[jnp.ndarray, Tuple[jnp.ndarray, PyTree]]:
+        if mutable is not None:
+            logger.warning("Usually when working with Transformers models `mutable` is None. Here it is not.")
```
[the remark this answers is not in the document]
Comment (albertogaspar) [reply fortuna/model/model_manager/classification.py:278]: Removed

File: fortuna/prob_model/posterior/sngp/transformers/auto_factory.py
[the remark this answers is not in the document]
Comment (albertogaspar) [reply fortuna/prob_model/posterior/sngp/transformers/auto_factory.py:1]: SNGP requires to change the model internally in order to add Spectral Normalization. I set it up in such a way that the API is the same as the one offered by HuggingFace. See the example below.

Pretrained bert model from HF:
```python
from transformers import FlaxAutoModelForSequenceClassification
model = FlaxAutoModelForSequenceClassification.from_pretrained("bert-base-cased")
```

Pretrained bert model to be trained via SNGP:
```python
from fortuna.prob_model.posterior.sngp.transformers import FlaxAutoSNGPModelForSequenceClassification
model = FlaxAutoSNGPModelForSequenceClassification.from_pretrained("bert-base-cased")
```
